SudokuSolver/SudokuSolver/CrossHatching/SudokuSolver.cs

```diff
@@ -0,0 +1,134 @@
+﻿using Microsoft.Toolkit.Extensions;
+using SudokuSolver.Common;
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace SudokuSolver.CrossHatching
+{
+    class SudokuSolver : BaseSudokuSolver
+    {
+        private Dictionary<int, List<(int i, int j)>> filledPos;
+        private Dictionary<int, int> remaining;
+        private Dictionary<int, Dictionary<int, List<int>>> graph;
+
+        public SudokuSolver() { }
+        public override Result<Sudoku> Solve(Sudoku sudoku)
+        {
+            BuildFilledPosAndRemaining(sudoku);
+            BuildGraph(sudoku);
+
+            remaining = remaining.OrderBy(kvpair => kvpair.Value).ToDictionary(x => x.Key, x => x.Value);
+
+            var remainingKeys = remaining.Keys.ToList();
+            int currentRemKey = 0;
+            int currentRow = 0;
+
+            Stopwatch sw = new Stopwatch();
+            sw.Start();
+            var result = TrySolve(
+                sudoku,
+                currentRemKey,
+                remainingKeys,
+                currentRow,
+                graph[remainingKeys[currentRemKey]].Keys.ToList()
+            );
+            sw.Stop();
+            Console.WriteLine(sw.Elapsed);
+
+            return result;
+        }
+
+        private Result<Sudoku> TrySolve(Sudoku sudoku, int currentRemKey, List<int> remainingKeys, int currentRow, List<int> rows)
+        {
+            var k = remainingKeys[currentRemKey];
+            var r = rows[currentRow];
+
+            foreach (var c in graph[k][r])
+            {
+                if (sudoku.Get(r, c) > 0) continue;
+
+                if (IsLegal(sudoku, r, c, k))
+                {
+                    sudoku.Set(r, c, k);
+
+                    if (currentRow < rows.Count - 1)
+                    {
+                        if (TrySolve(sudoku, currentRemKey, remainingKeys, currentRow + 1, rows).Success) return Result<Sudoku>.Ok(sudoku);
+
+                        sudoku.Set(r, c, 0);
+                        continue;
+                    }
+                    else
+                    {
+                        if(currentRemKey >= remainingKeys.Count - 1) return Result<Sudoku>.Ok(sudoku);
+
+                        if (TrySolve(sudoku, currentRemKey + 1, remainingKeys, 0, graph[remainingKeys[currentRemKey + 1]].Keys.ToList()).Success) return Result<Sudoku>.Ok(sudoku);
+
+                        sudoku.Set(r, c, 0);
+                        continue;
+                    }
+                }
+            }
+
+            return Result<Sudoku>.Fail(sudoku, "No solution exists");
+        }
+
+        private void BuildFilledPosAndRemaining(Sudoku sudoku)
+        {
+            filledPos = new Dictionary<int, List<(int i, int j)>>();
+            remaining = new Dictionary<int, int>();
```

Comment:
Można to zainicjalizować albo w konstruktorze, albo bezpośrednio przy definicji pola w klasie, oczywiście przy założeniu że ten Solver to będzie instancja na jednorazową kalkulację (trochę jak serwisy) do sudoku a nie na wielokrotne użycie, jeśli masz w głowie wielokrotne użycie to takie podejście jest ok.

Reply:
Taki był zamysł właśnie, by było to wielokrotne użycie.